expat/tests/misc_tests.c

```diff
@@ -406,6 +406,44 @@ START_TEST(test_misc_create_external_entity_parser_with_null_context) {
 }
 END_TEST
 
+START_TEST(test_misc_general_entities_support) {
+  const char *const doc
+      = "<!DOCTYPE r [\n"
+        "<!ENTITY e1 'v1'>\n"
+        "<!ENTITY e2 SYSTEM 'v2'>\n"
+        "]>\n"
+        "<r a1='[&e1;]'>[&e1;][&e2;][&amp;&apos;&gt;&lt;&quot;]</r>";
+
+  CharData storage;
+  CharData_Init(&storage);
+
+  XML_Parser parser = XML_ParserCreate(NULL);
+  XML_SetUserData(parser, &storage);
+  XML_SetStartElementHandler(parser, accumulate_start_element);
+  XML_SetExternalEntityRefHandler(parser,
+                                  external_entity_failer__if_not_xml_ge);
+  XML_SetEntityDeclHandler(parser, accumulate_entity_decl);
+  XML_SetCharacterDataHandler(parser, accumulate_char_data);
+
+  if (_XML_Parse_SINGLE_BYTES(parser, doc, (int)strlen(doc), XML_TRUE)
+      != XML_STATUS_OK) {
+    xml_failure(parser);
+  }
+
+  XML_ParserFree(parser);
+
+  CharData_CheckXMLChars(&storage,
+#if XML_GE == 1
+                         XCS("e1=v1\n") XCS("e2=(null)\n") XCS("(r(a1=[v1]))\n")
+                             XCS("[v1][][&'><\"]")
+#else
+                         XCS("e1=&amp;e1;\n") XCS("e2=(null)\n")
```

Comment:
The `&amp;` here surprised me, since I was expecting that the app would only see "clean" text... but maybe there's good reason for an EntityDeclHandler to get the "escaped" variant. In any case, I don't think it's related to this pull request.

Reply:
The handler gets the original replacement text, yes, and because the replacement text is parsed after replacement, it needs `&amp;` rather than plain `&` to end up with text `&e1;` eventually. I'll mark this as resolved (because that's my impression) by happy to open this box up again as needed.